enterprise/backend/src/metabase_enterprise/semantic_search/embedding.clj

```diff
@@ -83,159 +81,140 @@
+(defmethod pull-model           "ollama" [{:keys [model-name]}]      (ollama-pull-model model-name))
+
+;;;; OpenAI impl
+
+(defn- openai-get-embeddings-batch [model-name texts]
+  (let [api-key (semantic-settings/openai-api-key)
+        endpoint "https://api.openai.com/v1/embeddings"]
+    (when-not api-key
+      (throw (ex-info "OpenAI API key not configured" {:setting "ee-openai-api-key"})))
     (try
-      (log/debug "Generating Ollama embedding for text of length:" (count text))
-      (-> (http/post "http://localhost:11434/api/embeddings"
-                     {:headers {"Content-Type" "application/json"}
-                      :body    (json/encode {:model (get-model)
-                                             :prompt text})})
+      (log/debug "Generating" (count texts) "OpenAI embeddings in batch")
+      (-> (http/post endpoint
+                     {:headers {"Content-Type" "application/json"
+                                "Authorization" (str "Bearer " api-key)}
+                      :body    (json/encode {:model model-name
+                                             :input texts})})
           :body
           (json/decode true)
-          :embedding)
+          :data
+          (->> (map :embedding)
+               vec))
       (catch Exception e
-        (log/error e "Failed to generate Ollama embedding for text of length:" (count text))
-        (throw e))))
+        (log/error e "Failed to generate OpenAI embeddings for batch of" (count texts) "texts"
+                   "with token count:" (count-tokens-batch texts))
+        (throw e)))))
 
-  (-get-embeddings-batch [this texts]
-    ;; Ollama doesn't have a native batch API, so we fall back to individual calls
-    ;; No special batching needed for Ollama - just process all texts
-    (log/debug "Generating" (count texts) "Ollama embeddings (using individual calls)")
-    (mapv #(-get-embedding this %) texts))
+(defn- openai-get-embedding [model-name text]
+  (try
+    (log/debug "Generating OpenAI embedding for text of length:" (count text))
+    (first (openai-get-embeddings-batch model-name text))
+    (catch Exception e
+      (log/error e "Failed to generate OpenAI embedding for text of length:" (count text))
+      (throw e))))
 
-  (-pull-model [_]
-    (try
-      (log/debug "Pulling embedding model from Ollama...")
-      (http/post "http://localhost:11434/api/pull"
-                 {:headers {"Content-Type" "application/json"}
-                  :body    (json/encode {:model (get-model)})})
-      (catch Exception e
-        (log/error e "Failed to pull embedding model")
-        (throw e))))
+(defmethod get-embedding        "openai" [{:keys [model-name]} text] (openai-get-embedding model-name text))
+(defmethod get-embeddings-batch "openai" [{:keys [model-name]} text] (openai-get-embeddings-batch model-name text))
+(defmethod pull-model           "openai" [_] (log/info "OpenAI provider does not require pulling a model"))
```

Comment:
bikeshed nit, feel free to disregard: it seems like dispatching on keywords rather than strings is the more common pattern in metabase (or at least the parts I'm familiar with drivers/lib). I don't know that there's any strong argument for or against here other than consistency, just a style nit to consider (unless there's a reason to prefer strings in which case, by all means :)

Reply:
To motivate my decision here with an anti-nit: you see I like to go with the grain of these things being defined in envars and in the (later given my ulterior motives) database as VARCHAR or TEXT. I find then there is less calling of `keyword` or `name` and suchlike. As they were unamespaced the keyword does not really add a great deal beyond 'looks different', the performance arguments etc I think do not really apply in this case.

I have nothing against keywords, I'm just old and grumpy!

>  the sage will go around obstacles rather than aggressively confront them – just like water. Lao Tzu (probably)

Of course, I can change them if there is going to be a metariot 👍